cmd/agent/subcommands/snmp/command.go

```diff
@@ -29,272 +30,267 @@ import (
+	}
+
+	if version != gosnmp.Version3 && connParams.CommunityString == "" {
+		// Set default community string if version 1 or 2c and no given community string
+		connParams.CommunityString = defaultCommunityString
 	}
 
+	// Authentication check
+	if version == gosnmp.Version3 && connParams.Username == "" {
+		return nil, fmt.Errorf("username is required for snmp v3")
+	}
+
+	port := connParams.Port
+	if port == 0 {
+		port = defaultPort
+	}
+
+	securityParams := &gosnmp.UsmSecurityParameters{}
+	var msgFlags gosnmp.SnmpV3MsgFlags
 	// Set v3 security parameters
-	if setVersion == gosnmp.Version3 {
+	if version == gosnmp.Version3 {
+		securityParams.UserName = connParams.Username
+		securityParams.AuthenticationPassphrase = connParams.AuthKey
+		securityParams.PrivacyPassphrase = connParams.PrivKey
+
 		// Authentication Protocol
-		switch strings.ToLower(cliParams.authProt) {
+		switch strings.ToLower(connParams.AuthProtocol) {
 		case "":
-			authProtocol = gosnmp.NoAuth
+			securityParams.AuthenticationProtocol = gosnmp.NoAuth
 		case "md5":
-			authProtocol = gosnmp.MD5
+			securityParams.AuthenticationProtocol = gosnmp.MD5
 		case "sha":
-			authProtocol = gosnmp.SHA
+			securityParams.AuthenticationProtocol = gosnmp.SHA
 		case "sha224", "sha-224":
-			authProtocol = gosnmp.SHA224
+			securityParams.AuthenticationProtocol = gosnmp.SHA224
 		case "sha256", "sha-256":
-			authProtocol = gosnmp.SHA256
+			securityParams.AuthenticationProtocol = gosnmp.SHA256
 		case "sha384", "sha-384":
-			authProtocol = gosnmp.SHA384
+			securityParams.AuthenticationProtocol = gosnmp.SHA384
 		case "sha512", "sha-512":
-			authProtocol = gosnmp.SHA512
+			securityParams.AuthenticationProtocol = gosnmp.SHA512
 		default:
-			fmt.Printf("Unsupported authentication protocol: %s \n", cliParams.authProt)
-			cliParams.cmd.Help() //nolint:errcheck
-			os.Exit(1)
-			return nil
+			return nil, fmt.Errorf("unsupported authentication protocol: %s", connParams.AuthProtocol)
 		}
 
 		// Privacy Protocol
-		switch strings.ToLower(cliParams.privProt) {
+		switch strings.ToLower(connParams.PrivProtocol) {
 		case "":
-			privProtocol = gosnmp.NoPriv
+			securityParams.PrivacyProtocol = gosnmp.NoPriv
 		case "des":
-			privProtocol = gosnmp.DES
+			securityParams.PrivacyProtocol = gosnmp.DES
 		case "aes":
-			privProtocol = gosnmp.AES
+			securityParams.PrivacyProtocol = gosnmp.AES
 		case "aes192":
-			privProtocol = gosnmp.AES192
+			securityParams.PrivacyProtocol = gosnmp.AES192
 		case "aes192c":
-			privProtocol = gosnmp.AES192C
+			securityParams.PrivacyProtocol = gosnmp.AES192C
 		case "aes256":
-			privProtocol = gosnmp.AES256
+			securityParams.PrivacyProtocol = gosnmp.AES256
 		case "aes256c":
-			privProtocol = gosnmp.AES256C
+			securityParams.PrivacyProtocol = gosnmp.AES256C
 		default:
-			fmt.Printf("Unsupported privacy protocol: %s \n", cliParams.privProt)
-			cliParams.cmd.Help() //nolint:errcheck
-			os.Exit(1)
```

Comment:
We actually want to keep this behaviour, if we have unsupported values we want to show the help that lists all the supported values for the different parameters and exit

Reply:
I think printing the full help text makes it hard to spot the actual error message. I've just refactored this again to use a custom Value type for the flags with values so that we get nicely formatted errors from cobra, which I think should solve this problem - now the error message looks like:
```sh
$ agent snmp walk localhost -x foo
Error: invalid argument "foo" for "-x, --priv-protocol" flag: must be one of DES|AES|AES192|AES192C|AES256|AES256C
```

Reply:
much better ! thanks for the change